course_info/icommons.py

```diff
@@ -1,80 +1,94 @@
-                                        extra_headers=self.headers,
-                                        timeout=60,
-                                        serialize=True,
-                                        trailing_slash=False)
+                                          extra_headers=self.headers,
+                                          timeout=60,
+                                          serialize=True,
+                                          trailing_slash=False)
 
     def get_course_info(self, course_instance_id):
-        course_info = {}
-        # get the course_instance data
-        try:
-            relative_url = '/course_instances/%s/?format=json' % course_instance_id
-            response = self.make_request('GET', relative_url, headers=self.headers)
-            course_info = response.data.copy()
-        except Exception as e:
-            log.exception(e.message)
+        cache_key = CACHE_KEY_COURSE_BY_COURSE_INSTANCE_ID.format(course_instance_id)
+        course_info = cache.get(cache_key)
+        if course_info is None:
+            course_info = {}
+            # get the course_instance data
+            try:
+                relative_url = '/course_instances/%s/?format=json' % course_instance_id
+                response = self.make_request('GET', relative_url,
+                                             headers=self.headers)
+                course_info = response.data.copy()
+                log_msg = "Caching course info for course_instance_id {}: {}"
+                log.debug(log_msg.format(course_instance_id, json.dumps(course_info)))
+                cache.set(cache_key, course_info)
+            except Exception as e:
+                log.exception(e.message)
         return course_info
 
     def get_course_info_by_canvas_course_id(self, canvas_course_id):
-        course_info = {}
-        # get the course_instance data
-        try:
-            relative_url = '/course_instances/?format=json&canvas_course_id=%s' % canvas_course_id
-            response = self.make_request('GET', relative_url, headers=self.headers)
-            results = response.data.get('results', [])
-            if len(results):
-                course_info = self.get_course_info(results[0]['course_instance_id'])
-        except Exception as e:
-            log.exception(e.message)
+        cache_key = CACHE_KEY_COURSE_BY_CANVAS_COURSE_ID.format(canvas_course_id)
+        course_info = cache.get(cache_key)
+        if course_info is None:
+            course_info = {}
+            # get the course_instance data
+            try:
+                relative_url = '/course_instances/?format=json&canvas_course_id=%s' % canvas_course_id
+                response = self.make_request('GET', relative_url,
+                                             headers=self.headers)
+                results = response.data.get('results', [])
+                if len(results):
+                    course_info = self.get_course_info(
```

Comment:
I know this is just the existing behavior, but the primary for an xlist may not actually be the first item in the list.  I _think_ we return xlist details from this endpoint, but not sure.